[quant] Adding quantized mul kernel

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24873 [quant] Rewrite of the scalar mul
* #24447 [quant] Adding Scalar add/mul.
* **#24444 [quant] Adding quantized mul kernel**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24444

Differential Revision: [D16844824](https://our.internmc.facebook.com/intern/diff/D16844824)

Commit:
Adding quantized mul kernel

Commit:
Update on "Adding quantized mul kernel"

Commit:
Update on "Adding quantized mul kernel"

Commit:
Update on "Adding quantized mul kernel"


Differential Revision: [D16844824](https://our.internmc.facebook.com/intern/diff/D16844824)

llyfacebook commented on torch/nn/quantized/modules/functional_modules.py:
```diff
@@ -93,7 +93,8 @@ def add(self, x, y):
     r"""Operation equivalent to ``torch.ops.quantized.mul``"""
     def mul(self, x, y):
-        raise NotImplementedError("Implementation of 'mul' is in progress...")
+        return ops.quantized.add(x, y, scale=self.scale,
```
ops.quantized.mul?

Commit:
Update on "Adding quantized mul kernel"


Differential Revision: [D16844824](https://our.internmc.facebook.com/intern/diff/D16844824)

Commit:
Update on "Adding quantized mul kernel"


Differential Revision: [D16844824](https://our.internmc.facebook.com/intern/diff/D16844824)

Commit:
Update on "Adding quantized mul kernel"


Differential Revision: [D16844824](https://our.internmc.facebook.com/intern/diff/D16844824)

Commit:
Update on "Adding quantized mul kernel"


Differential Revision: [D16844824](https://our.internmc.facebook.com/intern/diff/D16844824)

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -0,0 +1,91 @@
+
+inline void check_inputs(const Tensor& qa, const Tensor& qb) {
+  TORCH_CHECK(qa.qscheme() == kPerTensorAffine ||
+              qa.qscheme() == kPerTensorSymmetric,
```
we won't have kPerTensorSymmetric anymore, you can remove this one.

jerryzh168 reviewed (approved):
LGTM

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -0,0 +1,91 @@
+  auto iter = TensorIterator::binary_op(out, self, other);
+  AT_DISPATCH_QINT_TYPES(out.scalar_type(), "qmul", [&]() {
+    cpu_kernel(iter, [&](scalar_t a, scalar_t b) -> scalar_t {
+      const auto da = at::dequantize_val(self_scale, self_zero_point, a);
```
don't we have a requantize_val API?

zafartahirov replied:
I was afraid of the error contribution. I will try it out!

zafartahirov replied:
Actually, I don't think that will help at all. The problem is that the multiplication doen't really work well with quantized types. My rationale is that if we requantize the values, we still need to compute `qc = round(a*b/s + z)`, where a, b are floating point multiplicands, and s, z are the output params. We are given `qa = round(a/s + z), qb = round(b/s + z)`, which are the requantized values. We are given `qa, qb` and need to compute `qc` I don't see a way on how to do it with out dequantizing them.

jerryzh168 replied:
OK, also noticed you are using const here, Tensor is a pointer so const is not necessary.

zafartahirov replied:
I think this is a `scalar_t` value, not tensor

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -0,0 +1,91 @@
+inline void check_inputs(const Tensor& qa, const Tensor& qb) {
+  TORCH_CHECK(qa.qscheme() == kPerTensorAffine ||
+              qa.qscheme() == kPerTensorSymmetric,
+              "Only per tensor quantization is suuported in Mul.");
```
NIT: typo

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -0,0 +1,91 @@
+              "Only per tensor quantization is suuported in Mul.");
+  TORCH_CHECK(qa.qscheme() == qb.qscheme(),
+              "Both inputs to Mul must have the same quantization shceme.");
+  TORCH_CHECK(qa.numel() == qb.numel(),
```
Why do we not support broadcast? For float, this works:
```
z = torch.rand(3,4,5)
v = torch.rand(3,1,1)
q = torch.mul(z,v)
``` 
and we should aim to match the same.

zafartahirov replied:
This works because there is a broadcasting logic there. It is in the todo list to implement the broadcasting for the quantized ops, but not part of this PR.

Commit:
Update on "[quant] Adding quantized mul kernel"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24444

Differential Revision: [D16844824](https://our.internmc.facebook.com/intern/diff/D16844824)
